common/misc_utils.py

```
@@ -238,7 +238,18 @@ def pip_install_torch():
+    # onnxruntime-gpu (pinned in pyproject.toml) needs its CUDA major version to
+    # match torch's. Installing from PyPI's default index pulls whichever CUDA
+    # toolkit torch's latest release bundles, which can drift out of sync and
+    # leave onnxruntime unable to load CUDAExecutionProvider (silent CPU
+    # fallback). Default to a known-compatible CUDA 12.1 build; override via
+    # TORCH_CUDA_INDEX_URL (set to "" to fall back to the default index) if
+    # your environment needs a different index or CUDA minor version.
+    index_url = os.environ.get("TORCH_CUDA_INDEX_URL", "https://download.pytorch.org/whl/cu121")
+    cmd = [sys.executable, "-m", "pip", "install", *pkg_names]
+    if index_url:
+        cmd += ["--index-url", index_url]
+    subprocess.check_call(cmd)
```

Ruff (0.15.21)

[error] `subprocess` call: check for execution of untrusted input

(S603)